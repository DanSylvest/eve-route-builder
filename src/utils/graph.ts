type System = {
  security: number;
  neighbors: number[];
};

export class Graph {
  protected systems: { [index: number]: System } = {};
  protected additionalConnections: { [index: number]: number[] } = {};
  protected avoidSystems: Set<number> = new Set();

  addSystem(system: number, security: number, neighbors: number[] = []) {
    this.systems[system] = { neighbors, security };
  }

  addChain(from: number, to: number) {
    if (this.systems[from].neighbors.indexOf(to) === -1) {
      this.systems[from].neighbors.push(to);
    }

    if (this.systems[to].neighbors.indexOf(from) === -1) {
      this.systems[to].neighbors.push(from);
    }
  }

  checkSystem(system: number) {
    return !!this.systems[system];
  }

  addAdditionalChain(from: number, to: number) {
    if (!this.additionalConnections[from]) {
      this.additionalConnections[from] = [];
    }

    if (!this.additionalConnections[to]) {
      this.additionalConnections[to] = [];
    }

    if (this.additionalConnections[from].indexOf(to) === -1) {
      this.additionalConnections[from].push(to);
    }

    if (this.additionalConnections[to].indexOf(from) === -1) {
      this.additionalConnections[to].push(from);
    }
  }

  avoidSystem(system: number) {
    this.avoidSystems.add(system);
  }

  neighbors(system) {
    if (this.avoidSystems.has(system)) {
      return [];
    }

    const additionalNeighbors = this.additionalConnections[system];

    if (additionalNeighbors) {
      return [...this.systems[system].neighbors, ...additionalNeighbors];
    }

    return this.systems[system].neighbors;
  }

  security(system) {
    return this.systems[system].security;
  }

  copy() {
    const g = new Graph();
    g.systems = this.systems;
    return g;
  }

  mergeWith(connections: number[][]) {
    const g = this.copy();
    connections.forEach(([origin, target]) => g.addAdditionalChain(origin, target));
    return g;
  }
}
