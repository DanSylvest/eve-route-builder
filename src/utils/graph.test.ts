import { Graph } from './graph';

const addSystems = (graph: Graph) => {
  graph.addSystem(100, 1);
  graph.addSystem(101, 1);
  graph.addSystem(102, 1);
  graph.addSystem(1000, 1);
};

describe('graph', () => {
  let graph: Graph;

  beforeEach(() => {
    graph = new Graph();
  });

  it('add systems', () => {
    addSystems(graph);
    expect(graph).toMatchSnapshot();
  });

  it('add chains and check neighbors', () => {
    addSystems(graph);
    graph.addChain(100, 101);
    graph.addChain(100, 102);
    graph.addChain(101, 102);
    expect(graph.neighbors(100)).toEqual([101, 102]);
    expect(graph.neighbors(101)).toEqual([100, 102]);
    expect(graph.neighbors(102)).toEqual([100, 101]);
  });

  it('check system exists', () => {
    addSystems(graph);
    expect(graph.checkSystem(100)).toEqual(true);
    expect(graph.checkSystem(-1)).toEqual(false);
  });

  it('add additional chain', () => {
    addSystems(graph);
    graph.addChain(100, 101);
    graph.addAdditionalChain(100, 1000);
    expect(graph.neighbors(100)).toEqual([101, 1000]);
  });

  it('check security', () => {
    addSystems(graph);
    expect(graph.security(100)).toEqual(1);
  });

  it('should copy', () => {
    addSystems(graph);
    expect(graph.copy()).toMatchSnapshot();
  });

  it('merge with additional connections', () => {
    addSystems(graph);
    expect(graph.mergeWith([[100, 1000]])).toMatchSnapshot();
  });
});
