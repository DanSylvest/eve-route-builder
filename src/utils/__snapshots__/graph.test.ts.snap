// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph add systems 1`] = `
Graph {
  "additionalConnections": Object {},
  "systems": Object {
    "100": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "1000": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "101": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "102": Object {
      "neighbors": Array [],
      "security": 1,
    },
  },
}
`;

exports[`graph merge with additional connections 1`] = `
Graph {
  "additionalConnections": Object {
    "100": Array [
      1000,
    ],
    "1000": Array [
      100,
    ],
  },
  "systems": Object {
    "100": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "1000": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "101": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "102": Object {
      "neighbors": Array [],
      "security": 1,
    },
  },
}
`;

exports[`graph should copy 1`] = `
Graph {
  "additionalConnections": Object {},
  "systems": Object {
    "100": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "1000": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "101": Object {
      "neighbors": Array [],
      "security": 1,
    },
    "102": Object {
      "neighbors": Array [],
      "security": 1,
    },
  },
}
`;
